lib/python-indent.coffee

```diff
@@ -75,32 +78,58 @@ module.exports = PythonIndent =
+    # the "column" input is ignored...?
     indent = editor.buildIndentString(tabs, column=offset)
 
-    # Set the indent
-    editor.getBuffer().setTextInRange([[row, 0], [row, 0]], indent)
-
-  unindentOnOpeningDelimiter: (editor, row, previousLine) ->
-    # Get all preceding lines
-    lines = if row > 0 then editor.buffer.lines[0..row - 1] else []
-
-    # Loop in reverse through lines
-    for line, i in lines by -1
-
-      # Set indent for row after declaration of block
-      if PythonIndent.openingDelimiterIndentRegex.test line
-
-        # Indent one tab-level past declaration
-        indent = editor.indentationForBufferRow i
-        indent += 1 if previousLine.slice(-1) is ':'
-        editor.setIndentationForBufferRow row, indent
-
-        # Stop trying after success
-        break
-
-  indentHanging: (editor, row, previousLine) ->
-    # Indent at the current block level plus the setting amount (1 or 2)
-    indent = (editor.indentationForBufferRow row) + (atom.config.get 'python-indent.hangingIndentTabs')
-
-    # Set the indent
-    editor.setIndentationForBufferRow row, indent
+    # The range of text to replace with our indent
+    # will need to change this for hard tabs, especially tricky for when
+    # hard tabs have mixture of tabs + spaces, which they can judging from
+    # the editor.buildIndentString function
+    startRange = [row, 0]
+    stopRange = [row, editor.indentationForBufferRow(row) * tabLength]
+    editor.getBuffer().setTextInRange([startRange, stopRange], indent)
+
+
+  parseLines: (lines) ->
+    # bracketStack is a stack of [row, column] arrays indicating the
+    # location of the opening bracket (square, curly, or parentheses)
+    bracketStack = []
+    # if we are in a string, this tells us what character introduced the string
+    # i.e., did this string start with ' or with "?
+    stringDelimiter = []
+
+    # NOTE: this parsing will only be correct if the python code is well-formed
+    #       statements like "[0, (1, 2])" might break the parsing
+
+    # loop over each line, adding to each stack
+    for row in [0 .. lines.length-1]
+        line = lines[row]
+
+        # boolean, whether or not the current character is being escaped
+        isEscaped = false
+
+        for col in [0 .. line.length-1]
+            c = line[col]
+
+            # if stringDelimiter is set, then we are in a string
+            if stringDelimiter.length
+                if isEscaped
+                    isEscaped = false
+                else
+                    if c == stringDelimiter
+                        stringDelimiter = []
+                    if c == '\\'
+                        isEscaped = true
+            else
+                if c == '#'
+                    break
+                if c in '[({'
+                    bracketStack.push([row, col])
+                if c in '})]'
+                    bracketStack.pop()
+                if c in '\'"'
```

Comment:
Also, perhaps it would be better to put this in an `if...else if` control structure. That way this can be short-circuited once a match is found.

I actually saw similar logic to this in a Sublime Text plugin once.